### PR TITLE
[READY] Black Ops Gamemode + OBJECTIVES REWRITE

### PR DESCRIPTION
**Pull request by Iamgoofball**
This gamemode is a subtype of Nuclear, and was coded to get a copy of Metal Gear Solid V: The Phantom Pain.

In short, it's Nuclear without the Nuke.

The Black Ops have 10 objectives they need to accomplish.

They need to kill 3 people.

They need to steal 3 different types of machines, and all of them on the station.

They need to destroy 3 different types of machines, and all of them on the station.

They need to rub an emag on the AI upload console to fry it.

I'll probably add more objectives if you can come up with them.

I want the objectives to be sort of like Mallcrasher from Payday 2, where the Black Ops would be people showing up to cause PROPERTY DAMAGE and smash shit up, kill people, abduct some, and fuck with highly expensive equipment.

The round can end in 4 ways.
1. The Black Ops return to the shuttle with their captives, and use the console, and escape with all the objectives complete. That's a Major Victory for the Ops.
2. The Black Ops return to the shuttle and escape without finishing all of the objectives. This is a Minor Victory for the ops if they accomplished some of their objectives.
3. The crew escapes alive before the Ops can return to the shuttle and escape with the objectives completed. This is a Minor Victory for the crew.
4. The Black Ops are killed on station by the crew. This is a Major Victory for the crew.
# OBJECTIVES REWRITE

Objectives are now GENERATED INSTEAD OF BEING A BIG GODDAMN IF ELSE CHAIN

Important bits:

Certain antags can roll objectives they couldn't before, such as Protect on traitor, or Debrain + Escape Alone on Wizard, etc.

I'm debating on increasing the amount of objectives traitors get by default too, since it's much easier to do now.
# FULTONS: Going to be replaced when SPRITERS MAKE A NEW IDEA FOR IT

For now, they're usable by Black Ops for securing machines they need to steal.

R&D can also produce them, but they can't lock onto Syndicate Beacons.

I'm wondering how these will be used in the round by the crew. I can remove them from R&D if wanted.

### CODE REVIEW

**Comment by Steelpoint** [comment]
This is not actually a bad idea, shitty OP aside.

It would add unpredictability in encountering Nuke Ops since now you have no way of knowing initally if they are here to blow the station up or to assassinate and steal stuff.

This has my :+1: 

I'm interested in seeing how this goes.

**Comment by RemieRichards** [comment]
@Steelpoint I actually like that he's being transparent about how he's actually <i>getting paid</i> (in the form of MGS:PP) But he definitely could have formatted it better.

**Comment by PKPenguin321** [comment]
neat :+1:

**Comment by LanCartwright** [comment]
Add a destroy percentage as an objective.

**Comment by Aranclanos** [comment]
Its filled with copypasta

**Comment by tuypo1** [comment]
sounds like a good idea I might have a look at the code in the morning but its bedtime right now

File: code/game/gamemodes/nuclear/blackops.dm [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+		syndicate.current << "<span class='notice'>You are a [syndicate_name()] agent!</span>"
+	var/obj_count = 1
+	for(var/datum/objective/objective in blackops_objective_holder.objectives)
+		syndicate.current << "<B>Objective #[obj_count]</B>: [objective.explanation_text]"
+		obj_count++
+	return
+
+/datum/game_mode/nuclear/blackops/proc/prepare_blackops_leader(datum/mind/synd_mind)
+	var/leader_title = pick("Czar", "Boss", "Commander", "Chief", "Kingpin", "Director", "Overlord")
+	spawn(1)
```
Comment (duncathan) [inline code/game/gamemodes/nuclear/blackops.dm]: What

File: code/game/gamemodes/nuclear/blackops_objectives.dm [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+				return 1
+			return 0
+		return 0
+	return 0
+
+/datum/objective/ai_mag/update_explanation_text()
+	..()
+	if(target && target.current)
+		explanation_text = "Apply an cryptographic sequencer to an AI upload console."
+	else
```
Comment (duncathan) [inline code/game/gamemodes/nuclear/blackops_objectives.dm]: Well at least now we're admitting that emags are mandatory

Comment (Iamgoofball) [inline code/game/gamemodes/nuclear/blackops_objectives.dm]: Yep.

Comment (Supermichael777) [inline code/game/gamemodes/nuclear/blackops_objectives.dm]: if your making emags mandatory GIVE THEM AN EMAG!(or change it to a corupted upload board as you have never been able to emag an AI befor)
also does completing this objective have an effect on the AI (like filling it with ion laws)?

Comment (Iamgoofball) [inline code/game/gamemodes/nuclear/blackops_objectives.dm]: Yes. It fills it with ion laws and purges the AI.

**Comment by WJohn** [comment]
I don't think kidnapping people is ever a good objective. Capturing them alive is basically impossible and isn't fun for the person getting kidnapped to be restrained for like half an hour or more. Maybe replace it with a couple of random steal objectives?

**Comment by Iamgoofball** [comment]
@WJohn 
You don't have to capture them alive. It's primarily about having to get them back to the shuttle in one piece. This isn't ninja mode kidnap where it's "lol sit there 30 minutes" on CC with nothing to do".

Plus, if you get them back alive, chances are the other ops have the other kidnaps and the other objectives complete, and the round's about to end.

Or, if the other ops DON'T have the kidnaps and are heading back out to go finish the job, now's your chance to escape. The ops shuttle contains a full set of tools, and insulated gloves to boot! They're not even behind doors, only windows and windoors. Make an escape with that space suit sitting by the door and get back to the station, and form an escape plan with your knowledge of where the spaceship is compared to the ops.

**Comment by AnturK** [comment]
Not a fan of emag getting even more power in all gamemodes.

**Comment by Iamgoofball** [comment]
@AnturK Well, when you make a successful PR to nerf/remove the emag, call me, until then, I'm going to stop pretending it isn't required gear in every syndicate based gamemode.

File: code/game/gamemodes/nuclear/blackops.dm (hunk 1 above)
Comment (Iamgoofball) [inline code/game/gamemodes/nuclear/blackops.dm]: literally the next line, read the comments

**Comment by duncathan** [comment]
Seriously though, if you're making the emag mandatory, then make sure you give one to every operative by default.

**Comment by Iamgoofball** [comment]
I'm planning on it.

**Comment by CosmicScientist** [comment]
Surely metagamers would go "Blood red/Shadow black ops seen, dismantle the law upload and reassemble it in a maintenance cubby hole!" to deny full greentext? Maybe instead have it be the AI itself since that's on the outside of the map where the operatives are closer to them than literally everyone is to the upload?

I like the idea though and the minor victory condition.

**Comment by Iamgoofball** [comment]
I specifically did the upload because I want the ops to have to siege a central part of the station to accomplish this, or take over a department(Research).

They have the same gear as nuke ops, as to prevent meta.

**Comment by CosmicScientist** [comment]
Alrighty then, I hope testing sees it to be achievable because I do like this.

**Comment by Iamgoofball** [comment]
It'll be awhile because I'm bundling a rewrite of granting objectives with this because intense gamemode code autism detected and I must kill.

**Comment by Czarcastic** [comment]
Destroying certain objectives on the station would be neat.

Grav gen, RnD server, SMES cells, the PA, the distro tanks in atmos, telecomms...

**Comment by Kromgar** [comment]
Goofball doing disclosures on paid code? Holy shit.

I never thought goofball would be so ethical

**Comment by Iamgoofball** [comment]
@tunderchief this is how you reply to posts

**Comment by optimumtact** [comment]
you know, is the person who paid for this aware that it doesn't necessarily imply the PR would get accepted?

**Comment by Iamgoofball** [comment]
@optimumtact Don't worry, I've already talked with him. It's actually technically for some other SS13 /tg/ based codebase, I'm doing it here for code review etc.

**Comment by Iamgoofball** [comment]
https://www.dropbox.com/s/ca1ytvz100zg97p/fulton_recovery.webm?dl=0

sneak peek of new content

**Comment by optimumtact** [comment]
How far up does it go?

**Comment by Razharas** [comment]
The shit in the video looks real fancy

**Comment by Thunder12345** [comment]
>It goes up
>Up
There truly is no ceiling on SS13

**Comment by RemieRichards** [comment]
@optimumtact if he's done it like I told him to, it's ```pixel_z```, so it actually goes off screen

**Comment by Iamgoofball** [comment]
Yep

**Comment by CosmicScientist** [comment]
Oh that looks very amusing. I wish that kind of animation was already used in the game.

**Comment by optimumtact** [comment]
So you will see it cross the whole station if you fultoned someone at the bottom?

**Comment by Razharas** [comment]
@Iamgoofball 
You alive?

**Comment by Iamgoofball** [comment]
@Razharas yes i am, still working on this

File: code/game/gamemodes/nuclear/blackops_objectives.dm (hunk 2 above)
Comment (Shadowlight213) [inline code/game/gamemodes/nuclear/blackops_objectives.dm]: Wait. Does it purge the ion laws as well?

**Comment by Iamgoofball** [comment]
this is still in progress don't worry
i'm trying to get 509 memery to work with fultons to fix some really big workarounds

File: code/game/gamemodes/nuclear/blackops_items.dm [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+				M.Stun(16) // Keep them from moving during the duration of the extraction
+				M.SetWeakened(0) // Make them upright if possible
+				M.buckled = 0 // Unbuckle them to prevent anchoring problems
+			else
+				A.anchored = 1
+				A.density = 0
+			var/obj/effect/extraction_holder/holder_obj = new(A.loc)
+			holder_obj.appearance = A.appearance
+			A.loc = holder_obj
+			balloon = image(icon,"extraction_balloon")
```
Comment (RemieRichards) [inline code/game/gamemodes/nuclear/blackops_items.dm]: why is this here

File: code/game/gamemodes/nuclear/blackops_items.dm [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+		return
+	else
+		if(A.loc == user || A == user) // no extracting stuff you're holding in your hands/yourself
+			return
+		if(A.anchored)
+			return
+		user << "<span class='notice'>You start attaching the pack to [A]...</span>"
+		if(do_after(user,50,target=A))
+			user << "<span class='notice'>You attach the pack to [A] and activate it.</span>"
+			var/image/balloon
```
Comment (RemieRichards) [inline code/game/gamemodes/nuclear/blackops_items.dm]: and not here?

File: code/game/gamemodes/nuclear/blackops_items.dm (hunk 3 above)
Comment (Iamgoofball) [inline code/game/gamemodes/nuclear/blackops_items.dm]: idfk, the image code will be cleaned up when 509 stops crashing on me

**Comment by Razharas** [comment]
@Iamgoofball 
How is it going

**Comment by Iamgoofball** [comment]
BYOND 509 should be mostly stable soonish and I can finish this then

**Comment by Iamgoofball** [comment]
# OH SHIT SON WE'RE READY FOR REVIEW

@Razharas @hornygranny @bs12-bot @Cheridan @RemieRichards @Shadowlight213 @MrPerson @MrStonedOne @paprka @paprika @god

**Comment by Iamgoofball** [comment]
Oh, and I'll add the beacon spawn points on the other 5 maps when this is merged.

File: code/game/gamemodes/antag_spawner.dm [hunk 5]
```diff
@@ -98,12 +98,9 @@
 	M.mind.name = newname
 	M.real_name = newname
 	M.name = newname
-	M.dna.update_dna_identity()
-	var/datum/objective/protect/new_objective = new /datum/objective/protect
-	new_objective.owner = M:mind
-	new_objective:target = usr:mind
+	var/datum/objective/default/protect/new_objective = add_objective(M.mind, /datum/objective/default/protect)
+	new_objective.target = usr:mind
```
Comment (duncathan) [inline code/game/gamemodes/antag_spawner.dm]: "I got rid of colons" he says

Comment (duncathan) [inline code/game/gamemodes/antag_spawner.dm]: Typecast usr you nerd

File: code/game/gamemodes/objective.dm [hunk 6]
```diff
@@ -6,6 +6,46 @@
...
+	objective.extra_prep()
+	objective.update_explanation_text()
+	if(announce_new_objectives)
+		if(M.special_role)
+			M.current << "<B>You are the [M.special_role].</B>"
+		var/obj_count = 1
+		for(var/datum/objective/OBJ in M.objectives)
+			M.current << "<B>Objective #[obj_count]</B>: [OBJ.explanation_text]"
+			obj_count++
+	return objective
```
Comment (MrPerson) [inline code/game/gamemodes/objective.dm]: I don't see how the add_objective(M, O) lines above aren't going to cause runtimes because O is an already-instanced objective.

File: code/game/gamemodes/nuclear/blackops_objectives.dm (hunk 2 above)
Comment (MrPerson) [inline code/game/gamemodes/nuclear/blackops_objectives.dm]: >AN CRYPTOGRAPHIC SEQUENCER

**Comment by MrPerson** [comment]
Lemme explain why blops is a bad idea real quick.

The reason modes like nuke and gang work is because both parties know exactly what the antag is trying to do. Blops isn't going to work because the crew has no idea what the blops are going to try to do, so how can they reasonably be expected to defend against them? Besides emag AI upload, which doesn't matter because they'll just rob tech storage to do it anyway. Save the unknown objectives for the hidden solo antags, please.

**Comment by Iamgoofball** [comment]
I'll separate black ops from the PR when J_Madison confirms he has the code lifted.

File: code/game/gamemodes/nuclear/blackops_items.dm [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+			if(istype(A, /mob/living))
+				var/mob/living/M = A
+				M.Stun(16) // Keep them from moving during the duration of the extraction
+				M.SetWeakened(0) // Make them upright if possible
+				M.buckled = 0 // Unbuckle them to prevent anchoring problems
```
Comment (AnturK) [inline code/game/gamemodes/nuclear/blackops_items.dm]:
```
if(M.buckled)
 M.buckled.unbuckle_mob()
```

File: code/game/gamemodes/nuclear/blackops_objectives.dm [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+		var/obj/machinery/M = new steal_machine
+		explanation_text = "Extract all [M]s located on the station."
+		qdel(M)
+	else
+		explanation_text = "Free Objective (Extract)"
+
+/datum/objective/extract_machine/check_completion()
+	var/total_machines
+	var/stolen_machines
+	for(var/obj/machinery/M in machines)
```
Comment (AnturK) [inline code/game/gamemodes/nuclear/blackops_objectives.dm]: No Z check ? Unless you want them to steal from mining and space

File: code/game/gamemodes/objective.dm [hunk 9]
```diff
@@ -6,6 +6,45 @@
 	var/completed = 0					//currently only used for custom objectives.
 	var/dangerrating = 0				//How hard the objective is, essentially. Used for dishing out objectives and checking overall victory.
 	var/martyr_compatible = 0			//If the objective is compatible with martyr objective, i.e. if you can still do it while dead.
+	var/required_role = null			//If the objective can only be done by a certain role, such as absorb for changelings.
+	var/randomgen = 1
+
+/proc/generate_objectives(var/datum/mind/M, var/objectives_made = 2, var/has_escape = TRUE)
```
Comment (AnturK) [inline code/game/gamemodes/objective.dm]: Why are these global procs instead of mind ones ?

Comment (Iamgoofball) [inline code/game/gamemodes/objective.dm]: I dunno, it works fine as-is.

File: code/game/gamemodes/nuclear/blackops_objectives.dm (hunk 8 above)
Comment (Iamgoofball) [inline code/game/gamemodes/nuclear/blackops_objectives.dm]: Whoops, good catch

File: code/game/gamemodes/nuclear/blackops_items.dm (hunk 7 above)
Comment (Iamgoofball) [inline code/game/gamemodes/nuclear/blackops_items.dm]: Thanks

**Comment by LanCartwright** [comment]
You're free to merge this.

**Comment by Iamgoofball** [comment]
Alrighty, I'll split up blops and objectives memery into separate PRs then

**Comment by MrStonedOne** [comment]
Please make a changelog,

You can add it to the PR's body rather then as a file now, see here: https://tgstation13.org/phpBB/viewtopic.php?f=5&t=5157

**Comment by Aranclanos** [comment]
almost 2 months since last commit, one month since last comment

**Comment by Iamgoofball** [comment]
whoops, forgot to close this when I moved objectives to its own PR, thanks aran
